File: merge_sorted_lists.rb
def merge_sorted_lists(left, right)
  result = []
  left_index, right_index = 0, 0

  loop do
    a = left[left_index]
    b = right[right_index]

    if a.nil? && b.nil?
      break
    elsif a.nil?
      result << b
      right_index += 1
    elsif b.nil?
      result << a
      left_index += 1
    else
      if a < b
        result << a
        left_index += 1
      else
        result << b
        right_index += 1
      end
    end
  end

  result
end



require 'minitest/autorun'

describe 'merger function' do

  it 'merges empty arrays' do
    actual = merge_sorted_lists [], []
    assert_equal [], actual
  end

  it 'merges with empty array' do
    actual = merge_sorted_lists [], [1, 2, 3]
    assert_equal [1, 2, 3], actual
  end

  it 'merges arrays of different length' do
    actual = merge_sorted_lists([1, 5], [2, 3, 4, 5, 6])
    assert_equal [1, 2, 3, 4, 5, 5, 6], actual
  end

end

File: merge_sorted_lists.go
package main

import (
	"fmt"
	"reflect"
)

func solution(left []int, right []int) []int {
	result := make([]int, len(left) + len(right))
	var (
		l = 0
		r = 0
		i = 0
	)

	for {
		if len(left) <= l && len(right) <= r {
			break
		} else if len(left) <= l {
			result[i] = right[r]
			r++
		} else if len(right) <= r {
			result[i] = left[l]
			l++
		} else {
			if left[l] < right[r] {
				result[i] = left[l]
				l++
			} else {
				result[i] = right[r]
				r++
			}
		}
		i++
	}

	return result
}

func main() {
	{
		expected := []int { }
		actual := solution([]int { }, []int { })
		fmt.Println("Merges empty arrays: ", reflect.DeepEqual(expected, actual))
	}

	{
		expected := []int { 1, 2, 3 }
		actual := solution([]int { }, []int { 1, 2, 3 })
		fmt.Println("Merges with empty array: ", reflect.DeepEqual(expected, actual))
	}

	{
		expected := []int { 1, 2, 3, 4, 5, 5, 6 }
		actual := solution([]int { 1, 5 }, []int { 2, 3, 4, 5, 6 })
		fmt.Println("Merges arrays of different length: ", reflect.DeepEqual(expected, actual))
	}
}
